mutation_partition_v2.cc

```diff
@@ -196,107 +150,163 @@ stop_iteration mutation_partition_v2::apply_monotonically(const schema& s, mutat
-        for (const range_tombstone_entry& rt : p._row_tombstones) {
-            position_in_partition_view pos = rt.position();
-            if (filtering_tombstones) {
-                if (cmp(res._pos, rt.end_position()) >= 0) {
-                    continue;
-                }
-                filtering_tombstones = false;
-                if (cmp(res._pos, rt.position()) > 0) {
-                    pos = res._pos;
-                }
+        // Pass only if continuity is the same on both sides and
+        // range tombstones for the intervals are the same on both sides (if intervals are continuous).
+        bool next_continuous = next_i == _rows.end() || next_i->continuous();
+        if (e.continuous() && next_continuous) {
+            tombstone next_range_tombstone = (next_i == _rows.end() ? tombstone{} : next_i->range_tombstone());
+            if (e.range_tombstone() != next_range_tombstone) {
+                return next_i;
             }
-            auto i = _rows.lower_bound(pos, cmp);
-            if (i == _rows.end()) {
-                break;
+        } else if (!e.continuous() && !next_continuous) {
+            if (!e.dummy() && e.range_tombstone()) {
+                return next_i;
             }
-            auto end = _rows.lower_bound(rt.end_position(), cmp);
+        } else {
+            return next_i;
+        }
 
-            auto tomb = _tombstone;
-            tomb.apply(rt.tombstone().tomb);
+        ++app_stats.rows_dropped_by_tombstones; // FIXME: it's more general than that now
 
-            if (apply_tombstone_to_rows(apply_resume::stage::range_tombstone_compaction, tomb, i, end) == stop_iteration::no) {
-                return stop_iteration::no;
-            }
+        i = _rows.erase(i);
+        if (tracker) {
+            tracker->remove(e);
         }
-    }
+        del(&e);
+        return next_i;
+    };
 
-    if (_row_tombstones.apply_monotonically(s, std::move(p._row_tombstones), preemptible) == stop_iteration::no) {
-        res = apply_resume::merging_range_tombstones();
-        return stop_iteration::no;
-    }
+    // Compacts a given row taking into account tombstones at all levels.
+    // Does not affect logical state of the partition.
+    auto compact = [&] (rows_entry& e) {
+        can_gc_fn never_gc = [](tombstone) { return false; };
+        ++app_stats.rows_compacted_with_tombstones;
+        e.row().compact_and_expire(s, _tombstone + e.range_tombstone(),
+             gc_clock::time_point::min(),  // no TTL expiration
+             never_gc,                     // no GC
+             gc_clock::time_point::min()); // no GC
+        // FIXME: Compact range tombstones
+    };
 
     if (p._tombstone) {
-        // p._tombstone is already applied to _tombstone
         rows_type::iterator i;
         if (res._stage == apply_resume::stage::partition_tombstone_compaction) {
-            i = _rows.lower_bound(res._pos, cmp);
+            i = _rows.upper_bound(res._pos, cmp);
```

Comment:
To ensure forward progress. The new code stores the position after compacting the row, before advancing the iterator. Old code was storing the iterator after advancing it. It's more robust against concurrent insertions since now we won't skip any interval if something was inserted after the last compacted entry. I don't think it was causing any real problems, it just seems better.